src/relations/s3_archival.py

```diff
@@ -0,0 +1,191 @@
+    """Client for s3:temporal relation."""
+
+    def __init__(self, charm):
+        """Construct.
+
+        Args:
+            charm: The charm to attach the hooks to.
+        """
+        super().__init__(charm, "s3")
+        self.charm = charm
+        charm.framework.observe(charm.s3_client.on.credentials_changed, self._on_s3_credentials_changed)
+        charm.framework.observe(charm.s3_client.on.credentials_gone, self._on_s3_credentials_gone)
+
+    @log_event_handler(logger)
+    def _on_s3_credentials_changed(self, event: CredentialsChangedEvent):
+        """Handle new s3:temporal relation.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        s3_parameters, missing_parameters = self._retrieve_s3_parameters()
+        if missing_parameters:
+            return
+
+        endpoint = _construct_endpoint(s3_parameters)
+        bucket_created = True
+
+        try:
+            _create_bucket_if_not_exists(s3_parameters, endpoint)
+        except (ClientError, ValueError):
+            bucket_created = False
+
+        self.charm._state.s3 = {
+            "bucket": s3_parameters.get("bucket"),
+            "endpoint": endpoint,
+            "region": s3_parameters.get("region"),
+            "aws_access_key_id": s3_parameters.get("access-key"),
+            "aws_secret_access_key": s3_parameters.get("secret-key"),
+            "uri_style": s3_parameters.get("s3-uri-style"),
+            "bucket_created": bucket_created,
+        }
+        self.charm._update(event)
+
+    @log_event_handler(logger)
+    def _on_s3_credentials_gone(self, event: CredentialsGoneEvent) -> None:
+        """Handle s3:temporal relation broken event.
+
+        Args:
+            event: The event triggered when the relation was broken.
+        """
+        if self.charm.unit.is_leader():
+            self.charm._state.s3 = None
+            self.charm._update(event)
+
+    def _retrieve_s3_parameters(self):
+        """Retrieve S3 parameters from the S3 integrator relation.
+
+        Returns:
+            s3 parameters and any missing parameters from the relation.
+        """
+        s3_parameters = self.charm.s3_client.get_s3_connection_info()
+        required_parameters = [
+            "bucket",
+            "access-key",
+            "secret-key",
+        ]
+        missing_required_parameters = [param for param in required_parameters if param not in s3_parameters]
+        if missing_required_parameters:
+            logger.warning(
+                f"Missing required S3 parameters in relation with S3 integrator: {missing_required_parameters}"
+            )
+            return {}, missing_required_parameters
+
+        # Add some sensible defaults (as expected by the code) for missing optional parameters
+        s3_parameters.setdefault("endpoint", "https://s3.amazonaws.com")
+        s3_parameters.setdefault("region")
```

Comment:
Is `None` the default that you want for `region`? it would be more clear to explicitly add this if so.